### PR TITLE
Optimize primitive writes in RecyclerBytesStreamOutput

### PR DESCRIPTION
**Pull request by original-brownbear**
We can inline the capacity check in these 2 spots as well, like we have for writing `long` etc. This gives a visible speedup in profiling benchmarks (we do quite a bit of single byte writing unfortunately).

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/common/io/stream/RecyclerBytesStreamOutput.java [hunk 1]
```diff
@@ -218,9 +230,7 @@ public BytesReference bytes() {
     }
 
     private void ensureCapacity(int bytesNeeded) {
-        if (bytesNeeded > pageSize - currentPageOffset) {
-            ensureCapacityFromPosition(position() + bytesNeeded);
-        }
+        ensureCapacityFromPosition(position() + bytesNeeded);
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/common/io/stream/RecyclerBytesStreamOutput.java:233]: Can we `assert bytesNeeded > pageSize - currentPageOffset` here?

**Review by DaveCTurner — APPROVED** [review]
LGTM (one nit) tho the title is misleading, we're not just touching the single-byte-write path.

**Comment by original-brownbear** [comment]
Thanks David!

File: server/src/main/java/org/elasticsearch/common/io/stream/RecyclerBytesStreamOutput.java (hunk 1 above)
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/common/io/stream/RecyclerBytesStreamOutput.java:258]: Done :)

**Pull request merged**